Detect Windows absolute paths on non-Windows CLI

thaJeztah opened the pull request:
Note that there is an alternative PR in https://github.com/docker/cli/pull/1871, but that PR changes the current behaviour, and might require more discussion. This PR tries to focus on just the basic problem.


addresses https://github.com/docker/cli/issues/1403
addresses https://github.com/moby/moby/issues/33746
fixes https://github.com/moby/moby/issues/34810


When deploying a stack using a relative path as bind-mount source in the compose file, the CLI converts the relative path to an absolute path, relative to the location of the docker-compose file.

This causes a problem when deploying a stack that uses an absolute Windows path, because a non Windows client will fail to detect that the path (e.g. `C:\somedir`) is an absolute path (and not a relative directory named `C:\`).

The existing code did already take Windows clients deploying a Linux stack into account (by checking if the path had a leading slash). This patch adds the reverse, and adds detection for Windows absolute paths on non-Windows clients.

The code used to detect Windows absolute paths is copied from the Golang filepath package;
https://github.com/golang/go/blob/1d0e94b1e13d5e8a323a63cd1cc1ef95290c9c36/src/path/filepath/path_windows.go#L12-L65

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
ping @silvin-lubecki @vdemeester @ddebroy PTAL

/cc @stevenfollis @david-yu @ivansharamok

thaJeztah commented:
Ah, booh! The linter doesn't like the Golang standard library;

```
cli/compose/loader/windows_path.go:35::warning: cyclomatic complexity 18 of function volumeNameLen() is high (> 16) (gocyclo)
```

thaJeztah commented on cli/compose/loader/windows_path.go:
```diff
@@ -0,0 +1,70 @@
+
+// volumeNameLen returns length of the leading volume name on Windows.
+// It returns 0 elsewhere.
+// nolint: gocyclo
```
Added this to silence the linter

ddebroy commented on cli/compose/loader/loader_test.go:
```diff
@@ -985,6 +985,84 @@ services:
+    image: mcr.microsoft.com/windows/servercore/iis:windowsservercore-ltsc2019
+    volumes:
+      - type: bind
+        source: c:\
```
Just to be safe, can the unit tests use `E:\` or `X:\` as the source path's drive (something other than `C:\`) in some of the cases?

thaJeztah replied:
Yes, I can change that; let me update

ddebroy commented on cli/compose/loader/windows_path.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+// isAbs reports whether the path is a Windows absolute path.
+func isAbs(filePath string) (b bool) {
```
should we have unit tests for `isAbs`?

thaJeztah replied:
All these were copied from the golang source code; I did have a quick peek if I could copy their tests as well, but couldn't find a test for these functions in isolation (at a glance)

thaJeztah replied:
oh; found where they live; I'll add some tests

thaJeztah commented:
Per discussion with @silvin-lubecki on slack: I removed the redundant `path.IsAbs` from the `isAbs()` function, and updated the test-file to also include the copyright from upstream

thaJeztah commented:
Booh; this one starts failing now all of a sudden; https://github.com/docker/cli/pull/1841#issuecomment-484855473

```
10:06:29 === FAIL: e2e/container TestSigProxyWithTTY (6.52s)
10:06:29     proxy_signal_test.go:38: terminating PID 3438
10:06:29     proxy_signal_test.go:42: timeout hit after 5s: expected status exited != running
```
